metals/src/main/scala/scala/meta/internal/metals/ReferenceProvider.scala

```diff
@@ -39,15 +44,17 @@ final class ReferenceProvider(
     definition: DefinitionProvider,
     trees: Trees,
     buildTargets: BuildTargets,
-) extends SemanticdbFeatureProvider {
-  private var referencedPackages: BloomFilter[CharSequence] =
-    BloomFilters.create(10000)
+    compilers: Compilers,
+    scalaVersionSelector: ScalaVersionSelector,
+)(implicit ec: ExecutionContext)
+    extends SemanticdbFeatureProvider {
 
   case class IndexEntry(
       id: BuildTargetIdentifier,
       bloom: BloomFilter[CharSequence],
   )
   val index: TrieMap[Path, IndexEntry] = TrieMap.empty
+  val tokenIndex: TrieMap[Path, IndexEntry] = TrieMap.empty
```

Comment:
Do you have an idea how much more memory this takes up? And maybe how much longer indexing takes? Or are the indexing times negligible since we already tokenize the files?

Reply:
I'll check both. Indexing should be similar but there may be additional cost to building the bloom filter.

Reply:
In metals repo:
`memory: identifier index using 1.37M (77,536 elements)`
Benchmarks:
```
Benchmark                                                   Mode  Cnt  Score   Error  Units
MetalsBench.alltoplevelsScalaIndex                            ss   10  0.419 ± 0.003   s/op
MetalsBench.alltoplevelsScalaIndexWithCollectIdents           ss   10  0.499 ± 0.011   s/op
MetalsBench.alltoplevelsScalaIndexWithBuildIdentifierIndex    ss   10  0.627 ± 0.016   s/op
```

Reply:
We should be fine then, that's not a lot :+1: 

The only issue I can think of is anything that is not named like apply methods, but we can try and handle that later.

Reply:
Sure. I guess the best way to go about it would be not to send `targetUri`s right away but have pc resolve the name. We could also use untyped trees. Anyway it will quickly get hairy for things like:
```
case class Foo() {
  def apply()
 }
 val f = Foo()
 val h = f()
 ```

Reply:
We could also look at synthetic symbols as we already have some logic for it (or at least the apply one) and we search for apply, it should be found properly